### ISSUE TITLE
Performance divergence between insert and update document

### ISSUE THREAD

**Issue by GavinMendelGleason** [terminusdb/terminusdb#693]
Via Seshat team:

Looking ahead to converting our code to read the full Equinox spreadsheet using the TerminusX schemes I modified my current test code to generate ‘large’ polities and time the system dealing with them. Specifically, there are 374 polities and ~47K facts in Equinox or about ~127 ‘facts’ per polity. So I modified the code to make up that number of facts by asserting 5 facts and 125 administrative level ‘facts' just to time the burden of updating a document with that many facts.

I then ran the code to incrementally assert those fake polities.  The initial polities take around 6s to transmit and commit using update_document(), which we would usually use rather than insert_document() since later versions of Equinox will replace the existing data for extand polities.  However, as the number of polities rise, so do the commit times, as they did in the older version of Terminus. E.g., by the time we get to the 50th polity the time has risen to 18s, 59 it takes 20s, etc. Roughly 0.28s increase per polity.  Worse, out of nowhere, an insert will take 100s of secs, e.g. 'Inserted af_durn_060 in 267.4s’  before reverting to previous times that continue to increase.  Quick estimate suggests it will require 6hrs to assert all the polities because of the increasing time, leaving aside the odd intermittent long pauses.  This behavior is reminiscent of the behavior from the previous version of Terminus.

Using insert_document() instead, which requires the polity to not exist in the DB, doesn’t exhibit the time increase or, apparently, the long pauses.  At ~5-10s/polity the total insert time dropped to a little over 40m.

So update_document()’s replacement/update takes time apparently to find the previous doc and do they replacement merge *even when there is no previous document to update* (so it should operate like insert_document()).  I think before we convert all our code to actually deal with real Equinox committed we should get this test code to operate at constant response using update_document(). 

```python
# A complete redo and clean up but still far from generating the schema from other data structures
from terminusdb_client.woqlschema import WOQLSchema, DocumentTemplate, LexicalKey, EnumTemplate, HashKey, RandomKey
from terminusdb_client import WOQLClient
from typing import List, Optional, Set
from enum import Enum
import pprint as pp
import json
import time
import sys
# CONTROL 
debug_values = False
n_polities = 374
bulk_insert = False
n_adm_levels = 125 # close to full Equinox burden
update_doc = False # false uses insert_documento

start_time = time.time()
# Your nickname from your new TerminusDB profile must be trhe user
# DEAD?? user = "benam"
# team = "Seshat"  # My team name.
team = "JSB"  # My team name.
### DON'T FORGET TO export TERMINUSDB_ACCESS_TOKEN into environment (check the link below)
### https://docs.terminusdb.com/v10.0/#/terminusx/get-your-api-key

endpoint = f"https://cloud.terminusdb.com/{team}/"
client = WOQLClient(endpoint)

#Connect to the db
# DEAD?? JSB client.connect(user=user, team=team, use_token=True)
client.connect(team=team, use_token=True)

exists = client.get_database("test_seshat")
if exists:
    ## BUG? This generates: UserWarning: Delete Database Warning: You have not specify the team, assuming JSB/test_seshat
    ## do we need to add test=team?  Yes. But why since it is already part of the connection?
    client.delete_database("test_seshat", team=team)
client.create_database("test_seshat")  # reset the DB
print('Connect execution time: %.1fs' % (time.time() - start_time))

start_time = time.time()

seshat_schema = WOQLSchema()

def dump_schema():
    global seshat_schema
    sso = seshat_schema.__dict__['object']
    for o in sso.values():
        pp.pprint(o.__dict__)


# Complete expedient since 3store doesn't yet permit Optional(Set(x)) or Optional(List(x))
# However, Set(x) is effectively Optional since the client (should) supplies {} as default, empty
# Nevertheless to experiment with Optional dates in SV we declare our own class which can be Optional
class _dateRange(DocumentTemplate):
    _schema = seshat_schema
    _subdocument = []
    _key = RandomKey() # All property instances that use SV need to be unique over the DB; this key is inherited properly
    # BUG: wanted to use 'from' and 'to' but 'from' causes a python 'invalid syntax' error
    from_date: int
    to_date: Optional[int ]

# NOTE: how come SV is not part of seshat_schema.__dict__['object']?? Because it didn't inherit from DocumentTemplate
# but to use we need to use ScopedValue before DocumentTemplate in the types below
class ScopedValue(DocumentTemplate):
    _schema = seshat_schema
    _subdocument = []
    _key = RandomKey() # All property instances that use SV need to be unique over the DB; this key is inherited properly
    #_key = LexicalKey("label") # not good enough
    # Everything in this class should be Optional

    # Eventually we will declare dates as follows and no longer need _dateRange
    # dates: Set[int ] # should be Optional like GYearRange below but we have to be explicit about low level types
    # BUG: You cannot have Optional[List[int ]]
    # terminusdb_client.errors.DatabaseError: Type error for json{'@class':"xsd:integer",'@type':"List"} which should be text
    # Eliminating Optional then requires always giving dates = [], not None
    # NOTE dates can be Set since it accepts empty, a singleton, or a distinct and ordered pair
    dates: Optional['_dateRange']
    # Confidence qualifiers
    # Semantics: if unknown is True then there will/must be no value set on the BoxedType tagged union
    # If unknown is True, suspected could be set True if it was supplied by an RA
    # and it would be removed (not set to False) once the expert approves
    # otherwise for a 'normal' fact it would have
    # 1) no unknown or suspected property,
    # 2) one of the BoxedType properties
    # 3) optionally one or both of disputed or inferred set
    unknown: Optional[bool]  # is the (typed) value explicitly unknown?
    # is the value (typically unknown) provided by an RA versus an expert
    suspected: Optional[bool]
    disputed: Optional[bool]  # is this one of several disputed values {}?
    inferred: Optional[bool]  # is the value inferred somehow?

# Type mixins (boxed classes) for property ScopedValues
# Each must have a value property that is required


class EpistemicStateEnum(EnumTemplate):
    '''Enumeration Type'''
    _schema = seshat_schema
    absent = ()
    absent_to_present = ()
    present = ()
    present_to_absent = ()

# ... define all the rest of the Seshat ScopedValue types from basic types and our enums
# The key is that these are the only types allowed on Polities, etc.
# They all inherit from ScopedValue
# They all provide the 'value' property of the proper type
# Some of these appear to be duplicates, e.g., GYearRange and IntegerRange, but we separate
# them because eventually they will have different constraints like ranges permitted, etc.

# BUG: gotta make value properties Optional to permit unknown declarations via SV
# However, making them all Optional yields (when asserting a value):
# terminusdb_client.errors.DatabaseError: Type error for json{'@class':"xsd:decimal",'@type':"Set"} which should be text
# I assume for the DecimalRange>

# BUG: If SV inherits from DocumentTemplate then all properties below complain about MRO problems
# TypeError: Cannot create a consistent method resolution
# order (MRO) for bases DocumentTemplate, ScopedValue
# REQUIREMENT: So order of sub classes is critical to have DocumentTemplate *last*!!

class EpistemicState(ScopedValue,DocumentTemplate):
    '''Present or absent'''
    _schema = seshat_schema
    value: Optional['EpistemicStateEnum']
    
class Integer(ScopedValue,DocumentTemplate):
    '''A simple number'''
    _schema = seshat_schema
    value: Optional[int ]

class IntegerRange(ScopedValue,DocumentTemplate):
    '''A simple number or range of integers'''
    _schema = seshat_schema
    # Can't be Optional yet
    value: Set[int ] # Set ensures order


class DecimalRange(ScopedValue,DocumentTemplate):
    '''Decimal number range'''
    _schema = seshat_schema
    # NEEDED?_subdocument = []
    # Can't be Optional yet
    value: Set[float ] # Set ensures order


class GYear(ScopedValue,DocumentTemplate):
    '''A particular Gregorian 4 digit year YYYY - negative years are BCE'''
    _schema = seshat_schema
    value: Optional[int ]


class GYearRange(ScopedValue,DocumentTemplate):
    '''A 4-digit Gregorian year, YYYY, or if uncertain, a range of years [YYYY,YYYY]'''
    _schema = seshat_schema
    # NEEDED? _subdocument = []
    # Can't be Optional yet
    value: Set[int ] # Set ensures order


class Boolean(ScopedValue,DocumentTemplate):
    _schema = seshat_schema
    # NEEDED? _subdocument = []
    value: Optional[bool ]

# Define example topic Section/Subsection classes (no properties on these)


class Topic(DocumentTemplate):
    '''Root of topic hierarchy'''
    _schema = seshat_schema
    _abstract = []


class GeneralInfo(Topic):
    '''General information variables'''
    _schema = seshat_schema
    _abstract = []


class SocialComplexity(Topic):
    '''Social Complexity variables'''
    _schema = seshat_schema
    _abstract = []


class Military(Topic):
   '''Miltiary variables'''
   _schema = seshat_schema
   _abstract = []


class Politics(Topic):
    '''Political variables'''
    _schema = seshat_schema
    _abstract = []
# A subsection


class Legal(Politics):  # Inherit from Politics as a subsection, not DocumentTemplate
    '''Legal system variables'''
    _schema = seshat_schema
    _abstract = []

# These are the properties on a Polity. There can be multiple Optional instances of each of these on a Polity
# Each inherits a boxed type class to hold the (scoped) value and a location in the Topic hierarchy for grouping
# Since these inherit from SV each instance will have a unique RandomKey() assigned
# NOTE: label will be used to parse from csv file strings to the proper class types and for pretty display on the website
# Note: our look-aside dicts need to maintain information that the 'Peak date' is the Variable from a csv file

class PeakDate(Integer,GeneralInfo):
    label= 'Peak date'
    _schema = seshat_schema
    _subdocument = []


class Duration(GYearRange, GeneralInfo):
    label = 'Duration'
    _schema = seshat_schema
    _subdocument = []


class Territory(DecimalRange,SocialComplexity):
    label = 'Polity territory'
    _schema = seshat_schema
    _subdocument = []


class ProfessionalMilitary(EpistemicState,Military):
    label = 'Professional military'
    _schema = seshat_schema
    _subdocument = []


class AdministrativeLevel(Integer,Politics):
    label = 'Administrative level'
    _schema = seshat_schema
    _subdocument = []


# TODO Need an example to inherit from Legal subsection

class Polity(DocumentTemplate):
    _schema = seshat_schema
    polid: str  # the equivalent of 'label'?  JSB? Note use of str, not 'xsd:string'?
    _key = LexicalKey(["polid"])
    originalID: str  # to compare with the old wiki polity page names (this could be Optional for new polities?? or dup polid?)
    # BUG: In addition to Optional(value) (see above) We need each of these properties to be an optional list of SVd instances of the right type
    # has to be list/set since we permit multiple contested facts and it has to be Optional because it is legit to no fact asserted
    # Why are they quoted strings?  because otherwise it returns the class object?
    # why doesn't Optional[List or Optional[Set work?  Errors when creating Polity -- They are fixing this.
    # terminusdb_client.errors.DatabaseError: Type error for json{'@class':"AdministrativeLevel",'@type':"List"} which should be text (or "Set")
    ## Move to Set, which permits eventually {} as a default.
    peak_date: Set['PeakDate']  # typically only one but could be disputed
    duration: Set['Duration']
    territory: Set['Territory']
    professional_military: Set['ProfessionalMilitary']
    admin_levels: Set['AdministrativeLevel']

seshat_schema.commit(client)
print('Schema execution time: %.1fs' % (time.time() - start_time))
# sys.exit(1)

if debug_values: # Dump the schema definition objects
    dump_schema()

## Now create some polities
start_time = time.time()

if bulk_insert:
    insert_type = 'bulk'
else:
    insert_type = 'incremental'
print('%s insert of %d polities with %d adminstrative levels using %s...' % (insert_type,n_polities,n_adm_levels,
                                                                             'update_document()' if update_doc else 'insert_document()'))
polities = []
for n in range(n_polities):
    name = 'af_durn_%03d' % n
    # name = 'af_durn_xxx'
    # We generate each polity with its own instances of values, not shared

    # Administrative levels: {0;1;...;2}:1800CE
    adm_lvls = set([AdministrativeLevel(value=lvl, dates= _dateRange(from_date=1800), disputed=True) for lvl in range(n_adm_levels)])
    pol = Polity(polid=name,
                 originalID='Afdurrn',
                 # Each entry in each property corresponds to a line in the csv file
                 # There are 7 entries so 7 lines
                 # peak_date = set(), # no data: nothing must be explicit set() since we can't have Optional Sets
                 # peak_date = {PeakDate(unknown=True, suspected=True)}, # Peak date: suspected unknown
                 peak_date={PeakDate(value = 1761)}, # no SV declarations Peak date: 1761 CE
                 duration={Duration(value = {1741, 1826})}, # Duration: 1741-1826 CE
                 territory={Territory(value = {6000, 8000}, dates = _dateRange(from_date=1772, to_date=1779)),# Polity territory: {6000-8000}:1772-1779CE; {17900-49000}:1800CE
                            Territory(value = {17900, 49000}, dates = _dateRange(from_date=1800))},
                 professional_military={ProfessionalMilitary(value=EpistemicStateEnum.present, inferred=True)}, # Professional military: inferred present
                 admin_levels=adm_lvls)
        
    # pp.pprint(pol.__dict__)
    if debug_values:
        print(pol.__dict__['polid'])
        for k,l in iter(pol.__dict__.items()):
            if type(l) is set or type(l) is list:
                print('%s:' % k)
                for v in l:
                    pp.pprint(v.__dict__)
    if bulk_insert:
        polities.append(pol)
    else:
        pol_start_time = time.time();
        # client.insert_document(pol,commit_msg='Inserting %s' % name)
        if update_doc:
            client.update_document(pol,commit_msg='Inserting %s' % name) # use update to wholesale define or update
        else:
            client.insert_document(pol,commit_msg='Inserting %s' % name) # use insert but assumes polity doesn't exist

        print('Inserted %s in %.1fs' % (name, (time.time() - pol_start_time)))
if bulk_insert:
    pol_start_time = time.time();
    if update_document:
        client.update_document(polities, commit_msg="Inserting all polities")
    else:
        client.insert_document(polities, commit_msg="Inserting all polities")
    print('Inserted up to %s in %.1fs' % (name, (time.time() - pol_start_time)))

elapsed_time = (time.time() - start_time)
time_per_polity = (elapsed_time/n_polities)
print('%s creation and insert execution time for %d polities: %.1fs (%.2fs/test polity)' % (insert_type,n_polities,elapsed_time, time_per_polity))

equinox_polities = 374.0
equiox_csv_lines = 47566.0
avg_lines_per_polity = equiox_csv_lines/equinox_polities # ~127
scale_factor = avg_lines_per_polity/(5+n_adm_levels) # there are nominally 7 lines in our example
time_per_polity *= scale_factor
print('Projected time for %s Equinox insert: %.2fs (%.2fs/full polity, %.2f scale)' % (insert_type, equinox_polities*time_per_polity, time_per_polity, scale_factor))

# Mar 7, 2021: previous 'incremental' (per polity) timings with a local server:
# Total assertions: 374 requiring inserts: 169681 deletes: 0  << why not 47K inserts?
# Total failed commits: 0
# Execution time: 2499.8s # ~6.6s/polity

# Oct 16, 2021 w/ remote server
# Projected time for bulk Equinox insert: 1485.26s (3.97s/full polity)
# Projected time for incremental Equinox insert: 9764.38s (26.11s/full polity)

```

**Comment by spl** [comment]
https://github.com/terminusdb/terminusdb-client-python/pull/256 is the Python change for this.

**Comment by spl** [comment]
This is fixed in the server and the Python client. We'll close this issue after the Python client is released with the change.

**Comment by spl** [comment]
Released with the client [`v10.0.23`](https://github.com/terminusdb/terminusdb-client-python/releases/tag/v10.0.23).
